gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java

```diff
@@ -0,0 +1,421 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.gcloud.spi;
+
+import static com.google.gcloud.spi.BigQueryRpc.Option.DELETE_CONTENTS;
+import static com.google.gcloud.spi.BigQueryRpc.Option.FIELDS;
+import static com.google.gcloud.spi.BigQueryRpc.Option.QUOTA_USER;
+import static com.google.gcloud.spi.BigQueryRpc.Option.START_INDEX;
+import static com.google.gcloud.spi.BigQueryRpc.Option.TIMEOUT;
+import static com.google.gcloud.spi.BigQueryRpc.Option.USER_IP;
+
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.jackson.JacksonFactory;
+import com.google.api.services.bigquery.Bigquery;
+import com.google.api.services.bigquery.model.Dataset;
+import com.google.api.services.bigquery.model.DatasetList;
+import com.google.api.services.bigquery.model.DatasetReference;
+import com.google.api.services.bigquery.model.GetQueryResultsResponse;
+import com.google.api.services.bigquery.model.Job;
+import com.google.api.services.bigquery.model.JobList;
+import com.google.api.services.bigquery.model.JobReference;
+import com.google.api.services.bigquery.model.QueryRequest;
+import com.google.api.services.bigquery.model.QueryResponse;
+import com.google.api.services.bigquery.model.Table;
+import com.google.api.services.bigquery.model.TableDataInsertAllRequest;
+import com.google.api.services.bigquery.model.TableDataInsertAllResponse;
+import com.google.api.services.bigquery.model.TableDataList;
+import com.google.api.services.bigquery.model.TableList;
+import com.google.api.services.bigquery.model.TableReference;
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+import static com.google.gcloud.spi.BigQueryRpc.Option.MAX_RESULTS;
+import static com.google.gcloud.spi.BigQueryRpc.Option.PAGE_TOKEN;
+
+import com.google.gcloud.bigquery.BigQueryException;
+import com.google.gcloud.bigquery.BigQueryOptions;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class DefaultBigQueryRpc implements BigQueryRpc {
+
+  public static final String DEFAULT_PROJECTION = "full";
+  // see: https://cloud.google.com/bigquery/troubleshooting-errors
+  private static final Set<Integer> RETRYABLE_CODES = ImmutableSet.of(503);
```

Comment:
Should we add [these](https://cloud.google.com/bigquery/loading-data-post-request#best-practices) other 500-level error codes?